clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -976,7 +1000,8 @@ mlir::LogicalResult CIRGenFunction::buildSwitchBody(
         builder.setInsertionPointToEnd(lastCaseBlock);
         res = buildStmt(c, /*useCurrentScope=*/!isa<CompoundStmt>(c));
       } else {
-        llvm_unreachable("statement doesn't belong to any case region, NYI");
+        checkCaseNoneStmt(*c);
```

Comment:
We should just call `buildStmt()` instead or rewrite more logic in a way that CIRGen just falls out naturally. Like I mentioned in previous reviews, this is dispatching another visitor logic just for the sake of grabbing information that could just be handled in our regular CIRGen visiting path.

Looking at `checkCaseNoneStmt` impl, specifically `Stmt::CaseStmtClass`/`DefaultStmtClass`: you should add a `buildCaseStmt` and `buildDefaultStmt` and call them from `buildSimpleStmt`. That code should already be part of CIRGen emission, and not something that does a side checking. If you need this info, you can walk LexScopes up to find if we are in a switch or not.

Reply:
If we call `buildStmt()`, for `ReturnStmt` we need to get `currLexScope->RetBlocks`, we expect the block is belong to the case region, but `CaseNoneStmt` has no region.
I'll keep working to find a solution. Pushed a `wip` commit once you want to have a look about the draft.